torch/distributed/_tools/mem_tracker.py

```diff
@@ -0,0 +1,811 @@
+                state_call = f"{state} # {i + 1}"
+                for dev, dev_snap in snapshot.items():
+                    if _rounding_fn(dev_snap[_TOTAL_KEY], divisor, 2) <= 0:
+                        continue
+                    row = {
+                        "State & Call": (
+                            state_call if state_call != last_state_call else ""
+                        ),
+                        "Device": str(dev),
+                    }
+                    last_state_call = state_call
+                    for k, v in dev_snap.items():
+                        row[f"{k}"] = f"{_rounding_fn(v, divisor, 2)} {units}"
+                    table_data.append(row)
+        print(tabulate(table_data, headers="keys", tablefmt="rst"))
+
+    except ImportError as err:
+        raise ImportError(
+            "Please install tabulate to use the tabulate option."
+        ) from err
+
+
+class MemTracker(TorchDispatchMode):
+    """
+    A TorchDispatchMode to track, categorize and attribute the tensor memory created or accessed within its context.
+
+    It categorizes the tracked tensors as parameters, buffers, activations, gradients, temporary memory and optimizer states
+    as defined by ``_MemRefType`` within its context. It captures memory `snapshots` for the modules, called within its context,
+    at various states defined by ``_ModState``.
+
+    Attributes:
+        memory_tracking: A weakref key dictionary to store the memory statistics of each module. Each key
+        is a reference to a module, and each value is a ``_ModMemStats`` object that stores the memory
+        statistics of the module.
+
+    Note:
+        The MemTracker should be used as a context manager. The modules, optimizers, and any other tensors created within
+        the context of MemTracker will be tracked by default. Any tensors or stateful objects such as modules, optimizers etc.
+        that need to be tracked but are created outside the MemTracker should be registered using the `track_external` method.
+        The `track_external` method should be called before the MemTracker is used. Any tensors created outside the ``MemTracker``
+        and not supplied to the `track_external` method will not be tracked by the ``MemTracker``.
+
+    Example usage:
+
+        .. code-block:: python
+
+            module = ...
+            optimizer = ...
+            inp = ...
+            mem_tracker = MemTracker()
+            mem_tracker.track_external(module, optimizer, inp)
+            with mem_tracker as mt:
+                loss = module(inp)
+                print("After Forward:")
+                mt.display_snapshot("current")
+                loss.backward()
+                optimizer.step()
+                optimizer.zero_grad()
+            mt.display_snapshot("peak")
+            mt.display_modulewise_snapshots(depth = 3, units = "MB")
+    """
+
+    def __init__(self) -> None:
+        self.memory_tracking = WeakIdKeyDictionary()
```

Comment:
Can we add an additional field to keep track of the execution order of the modules? For instance, 
* in `__init__`, add: 
```self.mods_exec_order = []```
* in `_pre_fw_hook`, after the check `if module not in self.memory_tracking:`, add:
```self.mods_exec_order.append(mod_name)```
* in `reset_mod_stats`, add:
```self.mods_exec_order.clear()```

Reply:
I can see the need, but that can be done with #128508. The only reason I am not adding that here is since there is no good way to display those stats. For instance, the pre-fw, post-fw, pre-bw and post-bw relative orders can be different for different modules. We will add it our run-time estimator.